pkg/pss/pss_test.go

```diff
@@ -27,202 +25,225 @@ func TestSend(t *testing.T) {
 
 	// send a message on topic1, check that only h1 is called
-	msg, err := trojan.NewMessage(topic1, payload)
+	chunk1, err := pss.Wrap(context.Background(), topic1, payload, recipient, targets)
 	if err != nil {
 		t.Fatal(err)
 	}
-	c, err := msg.Wrap(context.Background(), targets)
-	if err != nil {
-		t.Fatal(err)
-	}
-	err = pss.TryUnwrap(context.Background(), c)
+	err = p.TryUnwrap(context.Background(), chunk1)
 	if err != nil {
 		t.Fatal(err)
 	}
 
-	ensureCalls(t, &mtx, &h1Calls, 1)
-	ensureCalls(t, &mtx, &h2Calls, 0)
+	waitHandlerCallback(t, &msgChan, 1)
+
+	ensureCalls(t, "h1", &h1Calls, 1)
+	ensureCalls(t, "h2", &h2Calls, 0)
 
 	// register another topic handler on the same topic
-	cleanup := pss.Register(topic1, h3)
-	err = pss.TryUnwrap(context.Background(), c)
+	cleanup := p.Register(topic1, h3)
+	err = p.TryUnwrap(context.Background(), chunk1)
 	if err != nil {
 		t.Fatal(err)
 	}
 
-	ensureCalls(t, &mtx, &h1Calls, 2)
-	ensureCalls(t, &mtx, &h2Calls, 0)
-	ensureCalls(t, &mtx, &h3Calls, 1)
+	waitHandlerCallback(t, &msgChan, 2)
+
+	ensureCalls(t, "h1", &h1Calls, 2)
+	ensureCalls(t, "h2", &h2Calls, 0)
+	ensureCalls(t, "h3", &h3Calls, 1)
 
 	cleanup() // remove the last handler
 
-	err = pss.TryUnwrap(context.Background(), c)
+	err = p.TryUnwrap(context.Background(), chunk1)
 	if err != nil {
 		t.Fatal(err)
 	}
 
-	ensureCalls(t, &mtx, &h1Calls, 3)
-	ensureCalls(t, &mtx, &h2Calls, 0)
-	ensureCalls(t, &mtx, &h3Calls, 1)
+	waitHandlerCallback(t, &msgChan, 1)
+
+	ensureCalls(t, "h1", &h1Calls, 3)
+	ensureCalls(t, "h2", &h2Calls, 0)
+	ensureCalls(t, "h3", &h3Calls, 1)
 
-	msg, err = trojan.NewMessage(topic2, payload)
+	chunk2, err := pss.Wrap(context.Background(), topic2, payload, recipient, targets)
 	if err != nil {
 		t.Fatal(err)
 	}
-	c, err = msg.Wrap(context.Background(), targets)
+	err = p.TryUnwrap(context.Background(), chunk2)
 	if err != nil {
 		t.Fatal(err)
 	}
 
-	err = pss.TryUnwrap(context.Background(), c)
-	if err != nil {
-		t.Fatal(err)
-	}
+	waitHandlerCallback(t, &msgChan, 1)
 
-	ensureCalls(t, &mtx, &h1Calls, 3)
-	ensureCalls(t, &mtx, &h2Calls, 1)
-	ensureCalls(t, &mtx, &h3Calls, 1)
+	ensureCalls(t, "h1", &h1Calls, 3)
+	ensureCalls(t, "h2", &h2Calls, 1)
+	ensureCalls(t, "h3", &h3Calls, 1)
 }
 
-func ensureCalls(t *testing.T, mtx *sync.Mutex, calls *int, exp int) {
+func waitHandlerCallback(t *testing.T, msgChan *chan struct{}, count int) {
 	t.Helper()
 
-	for i := 0; i < 10; i++ {
-		mtx.Lock()
-		if *calls == exp {
-			mtx.Unlock()
-			return
+	for received := 0; received < count; received++ {
+		select {
+		case <-*msgChan:
+		case <-time.After(1 * time.Second):
+			t.Fatal("reached timeout while waiting for handler message")
 		}
-		mtx.Unlock()
-		<-time.After(100 * time.Millisecond)
 	}
-	t.Fatal("timed out waiting for value")
+}
+
+func ensureCalls(t *testing.T, name string, calls *int, exp int) {
+	t.Helper()
+
+	if exp != *calls {
+		t.Fatalf("%s expected %d calls, found %d", name, exp, *calls)
```

Comment:
adding the name of the handler that failed in not necessary, since `ensureCalls` calls `t.Helper()`, when failing it will show the line in which the failure occured on the calling context, so you would know exactly with which handler it was called and as of such you know which handler failed. so the string is not necessary. you will see that this pattern does not exist elsewhere in the codebase for this reason